### PR TITLE
Enable replaceWhere when DPO (dynamic partition overwrite) is enabled in the Spark configuration

### PR DESCRIPTION
**Pull request by allisonport-db**
### Description

This PR changes how `replaceWhere` and DPO interact with each other.

Current behavior:
- DPO + replaceWhere always throws an error

Behavior in this PR:
- DPO in spark session configuration + replaceWhere = data overwritten according to replaceWhere
- DPO as DataFrameWriter option + replaceWhere = throw error

### Testing

Updates and adds a test to `DeltaSuite`.

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala [hunk 1]
```diff
@@ -873,113 +904,106 @@ class DeltaSuite extends QueryTest
...
       }
     }
   }
 
-  test(
-    "batch write: append, dynamic partition overwrite conf not supported with replaceWhere") {
-    withSQLConf(DeltaSQLConf.DYNAMIC_PARTITION_OVERWRITE_ENABLED.key -> "true") {
-      withSQLConf(SQLConf.PARTITION_OVERWRITE_MODE.key -> "dynamic") {
-        withTempDir { tempDir =>
-          def data: DataFrame = spark.read.format("delta").load(tempDir.toString)
```
Comment (allisonport-db) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala:898]: this test is removed and replaced by the above two tests.

each following test name, and contents diffs are shifted by one.
I just fixed some style issues and since that messed with the indents the diff is messed up.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala [hunk 2]
```diff
@@ -873,113 +904,106 @@ class DeltaSuite extends QueryTest
...
+          .save(tempDir.getCanonicalPath)
+        checkDatasetUnorderly(data.select("value").as[Int], 1, 5)
       }
     }
   }
 
-  test(
-    "batch write: append, overwrite without partitions should ignore partition overwrite mode") {
+  test("batch write: append, overwrite non-partitioned table with replaceWhere ignores partition " +
+    "overwrite mode option") {
```
Comment (allisonport-db) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaSuite.scala:1004]: this is a new test. the diffs are messed up again, and `"batch write: append, overwrite without partitions should ignore partition overwrite mode"` is above... all the test(...) { are shifted down one with their contents

**Comment by allisonport-db** [comment]
hey @koertkuipers I opened this PR to implement the behavior we discussed on #371 if you want to take a look

**Comment by koertkuipers** [comment]
> hey @koertkuipers I opened this PR to implement the behavior we discussed on #371 if you want to take a look

looks great.
maybe also update the class scaladoc for WriteIntoDelta?
i dont think replaceWhere can be uses for SQL INSERT OVERWRITE right? so no additional tests needed for this?

**Comment by allisonport-db** [comment]
@koertkuipers yep, just for dataframe writes. I'll update the class doc thanks!

File: core/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala [hunk 3]
```diff
@@ -140,12 +140,16 @@ case class WriteIntoDelta(
       val useDynamic = txn.metadata.partitionColumns.nonEmpty &&
         options.isDynamicPartitionOverwriteMode
       options.replaceWhere.foreach { _ =>
-        if (useDynamic) {
-          throw new AnalysisException(s"'${DeltaOptions.REPLACE_WHERE_OPTION}' cannot be used" +
-            s" when '${DeltaOptions.PARTITION_OVERWRITE_MODE_OPTION}' is set to 'DYNAMIC'")
+        if (useDynamic && options.partitionOverwriteModeInOptions) {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/sql/delta/commands/WriteIntoDelta.scala:143]: this logic is a bit hard to follow as there is implicit assumption here that if partition overwrite mode is set in an option AND dynamic is effectively enabled (as a result of the complex code in DeltaOptions).. that means that dynamic must be enabled via the option. this is not obvious from the code.

i would write it such that its more obvious... like
```
if (isTablePartitioned && isDynamicPartitionOverwriteSetInOption && isReplaceWhereSetInOption)
```

its more verbose to create these fields .. but the results is more readable and obvious.
